### DISCUSSION TITLE
Nesting language highlights (e.g. rst in Python docstrings)

### DISCUSSION THREAD

**Discussion by BrunoMSantos** [nvim-treesitter/nvim-treesitter#4655] [Q&A]
Hey, I'm _really_ new to this treesitter thing and to the whole query bit, but I was hoping I could highlight docstrings in certain languages. I was especially hoping that this would give me better (not to mention easier) results than some clunky 'old vim' syntax configuration I had.

Notably, I was exploring if I could do rst in Python's docstrings and in C's / C++'s `/** */` block comments. In some jobs, doxygen inside C/C++ might also be nice, but I don't think there's a parser for that at the moment. Despite some early results I quickly found issues, which hopefully someone might be able to help with.

---------------------------------------------------------

Starting with Python, it seems pretty easy and in fact I _think_ I got it right by injecting the query:
```scm
(expression_statement (string (string_content) @rst))
```

The relevant comments do get highlighted as rst and it looks pretty much like what I'd want. There are two nagging issues I'm facing though: can't get the 'paragraph' node to be in the colour I want (still shows up in 'string' colour, other rst nodes are fine) and it causes an issue in `:checkhealth` for starlark.

To try and override the highlight, I tried this in the Python's `after/ftplugin` (colour unimportant, I was just looking for a change):
```lua
vim.api.nvim_set_hl(0, '@string.spell', { fg = 'white', bg = 'bg' })
```
But so far no change. If I do this instead, then it works, but all strings (and comments) get recoloured, which is not what I want:
```lua
vim.api.nvim_set_hl(0, '@spell', { fg = 'white', bg = 'bg' })
```

As for the starlark issue, I have no clue, but this is the error I get:
```
- ERROR: starlark(injections): /usr/share/nvim/runtime/lua/vim/treesitter/query.lua:219: query: invalid node type at position 216 for language starlark                                                             
  starlark(injections) is concatenated from the following files:                                                                                                                                                    
  |    [OK]:"/home/bruno/.local/share/nvim/lazy/nvim-treesitter/queries/python/injections.scm"                                                                                                                      
  | [ERROR]:"/home/bruno/.config/nvim/after/queries/python/injections.scm", failed to load: /usr/share/nvim/runtime/lua/vim/treesitter/query.lua:219: query: invalid node type at position 42 for language starlark 
  |    [OK]:"/home/bruno/.local/share/nvim/lazy/nvim-treesitter/queries/starlark/injections.scm"                                                                                                                    
```
I believe my injection file is getting roped in by starlark's inclusion of Python's injections. Is my injection query wrong? Or is this perhaps a bug? Otherwise, is there a way I can extend Python this way without the change affecting starlark as well?

---------------------------------------------------------------------------

Now, I'm not so hopeful about rst in C/C++, but I was looking to see if these type of comments were doable and go from there:
```c
/**
 * TITLE
 * =====
 *
 * .. note:: noted
 */
```

Even with no other considerations, I got stuck trying to filter out the `*`s. I was trying to do it with `offset`, but I clearly have no clue how to write this. Closest I got was this, which looks funny to me and indeed doesn't work:
```scm
((comment) @rst (#offset! @rst 1 3 -1 0))
```
Any directions would be welcome. And if this is entirely hopeless at the moment, that will be good to know too.

And yes, I know rst in C is not a thing, but one can [hope](https://github.com/jnikula/hawkmoth) :)

**Comment by ObserverOfTime** [comment]
See #4199 for rst in Python docstrings.

**Reply by theHamsta** [reply]
Please try https://github.com/neovim/neovim/pull/22784 or the query linter from nvim-treesitter/playground. Use the playground to get the correct name of the nodes for starklang.

I would avoid ftplugin to set filetype specific highlights. Use `:Inspect` to get the full highlight capture under cursor. It should have a `@string.python` suffix. You can use this to set ft-specifc highlights 
![image](https://user-images.githubusercontent.com/7189118/233791142-3db0a240-4204-4690-913c-b749dee2863a.png)

**Reply by BrunoMSantos** [reply]
> Please try [neovim/neovim#22784](https://github.com/neovim/neovim/pull/22784) or the query linter from nvim-treesitter/playground. Use the playground to get the correct name of the nodes for starklang.

I'll try to play around with that. However, I took `expression_statement.string_content` from the playground for Python, and the the query above selects exactly what I want in Python (even if poorly written perhaps). How would the starklang names help?

> I would avoid ftplugin to set filetype specific highlights. Use `:Inspect` to get the full highlight capture under cursor. It should have a `@string.python` suffix. You can use this to set ft-specifc highlights ![image](https://user-images.githubusercontent.com/7189118/233791142-3db0a240-4204-4690-913c-b749dee2863a.png)

Hmm, I had been using `:TSHighlightCapturesUnderCursor`, but indeed `:Inspect` seems to be what I wanted. For the record, I managed to affect only the parts I care about with `@spell.rst`, though I have to check for other implications as that might affect rst files themselves in unwanted ways.

Thanks for the tips!

**Reply by BrunoMSantos** [reply]
> See #4199 for rst in Python docstrings.

Thanks. I gave it a go, but I don't see any difference to what I get with my query to be honest. At least in the files I tested. I'll keep an eye on it though.

edit: I did find a difference in the end. Either way, I learnt something from it so, thanks again for pointing it out.
